File: Quick_Sort/Objective-C/QuickSort.m
//
//  QuickSort.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "QuickSort.h"

@implementation QuickSort

+ (NSArray<NSNumber *> *)sortedArray:(NSArray<NSNumber *> *)array order:(QuickSortOrder)order {
    NSUInteger arrayCount = array.count;

    if (arrayCount < 2) {
        return array.copy;
    }

    NSUInteger pivotIndex = arc4random() % arrayCount;
    NSNumber *pivotNumber = array[pivotIndex];

    NSMutableArray *lessArray = [NSMutableArray new];
    NSMutableArray *equalArray = [NSMutableArray new];
    NSMutableArray *greaterArray = [NSMutableArray new];

    for (NSNumber *number in array) {
        switch ([number compare:pivotNumber]) {
            case NSOrderedDescending:
                [lessArray addObject:number];
                break;
            case NSOrderedSame:
                [equalArray addObject:number];
                break;
            default:
                [greaterArray addObject:number];
                break;
        }
    }

    NSArray *lessSortedArray = [self sortedArray:lessArray order:order];
    NSArray *greaterSortedArray = [self sortedArray:greaterArray order:order];

    return [self mergedLessArray:lessSortedArray equalArray:equalArray greaterArray:greaterSortedArray order:order];
}

+ (NSArray *)mergedLessArray:(NSArray *)lessArray equalArray:(NSArray *)equalArray greaterArray:(NSArray *)greaterArray order:(QuickSortOrder)order {
    NSMutableArray *resultArray = [NSMutableArray arrayWithCapacity:lessArray.count + equalArray.count + greaterArray.count];

    if (order == QuickSortOrderAsc) {
        [resultArray addObjectsFromArray:greaterArray];
        [resultArray addObjectsFromArray:equalArray];
        [resultArray addObjectsFromArray:lessArray];
    }
    else {
        [resultArray addObjectsFromArray:lessArray];
        [resultArray addObjectsFromArray:equalArray];
        [resultArray addObjectsFromArray:greaterArray];
    }

    return resultArray.copy;
}

@end

File: Quick_Sort/Swift/QuickSort.swift
//
//  QuickSort.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct QuickSort {

    enum SortOrder {
        case ASC
        case DESC
    }

    /** Complexity is O(n * log(n)) */
    static public func sort<T: Comparable>(_ array: [T?]?, order: SortOrder) -> [T?]? {
        let arrayCount = array?.count ?? 0

        if arrayCount < 2 {
            return array
        }

        let pivotIndex = Int(arc4random()) % arrayCount
        let pivotNumber = array![pivotIndex]

        var lessArray: [T?] = []
        var equalArray: [T?] = []
        var greaterArray: [T?] = []

        for number in array! {
            if (number == nil && pivotNumber != nil) || (number != nil && pivotNumber != nil && number! < pivotNumber!) {
                lessArray.append(number)
            }
            else if (number == nil && pivotNumber == nil) || (number != nil && pivotNumber != nil && number! == pivotNumber!) {
                equalArray.append(number)
            }
            else {
                greaterArray.append(number)
            }
        }

        let lessSortedArray = sort(lessArray, order: order)
        let greaterSortedArray = sort(greaterArray, order: order)

        return merge(lessSortedArray!, equalArray, greaterSortedArray!, order: order)
    }

    static private func merge<T: Comparable>(_ lessArray: [T?], _ equalArray: [T?], _ greaterArray: [T?], order: SortOrder) -> [T?] {
        var resultArray: [T?] = []

        if order == .ASC {
            resultArray.append(contentsOf: lessArray)
            resultArray.append(contentsOf: equalArray)
            resultArray.append(contentsOf: greaterArray)
        }
        else {
            resultArray.append(contentsOf: greaterArray)
            resultArray.append(contentsOf: equalArray)
            resultArray.append(contentsOf: lessArray)
        }

        return resultArray
    }

}
